giraf/binaryGiraf/binReader.go

```diff
@@ -0,0 +1,269 @@
+package binaryGiraf
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"github.com/biogo/hts/bgzf"
+	"github.com/vertgenlab/gonomics/cigar"
+	"github.com/vertgenlab/gonomics/common"
+	"github.com/vertgenlab/gonomics/dna"
+	"github.com/vertgenlab/gonomics/dnaThreeBit"
+	"github.com/vertgenlab/gonomics/fileio"
+	"github.com/vertgenlab/gonomics/giraf"
+	"github.com/vertgenlab/gonomics/simpleGraph"
+	"io"
+	"log"
+	"strings"
+)
+
+// The BinReader struct wraps the bgzf reader from the biogo repository with a bytes buffer to store encoded giraf records
+type BinReader struct {
+	bg       *bgzf.Reader
+	currData *bytes.Buffer
+}
+
+// NewBinReader creates a new BinReader
+func NewBinReader(file io.Reader) *BinReader {
+	reader, err := bgzf.NewReader(file, 1) //TODO: Play with different levels of concurrency
+	common.ExitIfError(err)
+	return &BinReader{
+		bg:       reader,
+		currData: &bytes.Buffer{},
+	}
+}
+
+// DecompressGiraf will decode a binary giraf file (.giraf.fe) and output a giraf file (.giraf)
+func DecompressGiraf(filename string, graph *simpleGraph.SimpleGraph) {
+	// Initialize infile
+	infile := fileio.EasyOpen(filename)
+	defer infile.Close()
+	reader := NewBinReader(infile.BuffReader)
+	var err error
+
+	// Initialize outfile
+	outfile := fileio.EasyCreate(strings.TrimSuffix(filename, ".fe"))
+	defer outfile.Close()
+
+	// Read info until EOF
+	var curr giraf.Giraf
+	for curr, err = reader.Read(graph); err != io.EOF; curr, err = reader.Read(graph) {
+		common.ExitIfError(err)
+		giraf.WriteGirafToFileHandle(outfile, &curr)
+	}
+
+	// Close reader
+	err = reader.bg.Close()
+	common.ExitIfError(err)
+}
+
+// The Read method for the BinWriter struct decompresses a single giraf record and writes to file
+func (br *BinReader) Read(g *simpleGraph.SimpleGraph) (giraf.Giraf, error) {
+	var answer giraf.Giraf
+	var bytesRead int
+	var err error
+	var i, j, runLen uint16
+	var buffer [4]byte
+
+	// reset data buffer
+	br.currData.Reset()
+
+	// blockSize (uint32)
+	bytesRead, err = br.bg.Read(buffer[:4])
+	if err == io.EOF || bytesRead != 4 {
+		return answer, io.EOF
+	}
+	common.ExitIfError(err)
+	blockSize := int(binary.LittleEndian.Uint32(buffer[:4]))
+
+	// fill currData buffer with blockSize of bytes
+	data := make([]byte, blockSize) // read giraf to intermediate buffer //TODO: this is not great, find a way to not make an intermediate buffer
```

Comment:
One suggestion is to have this intermediate buffer as a struct field inside BinReader{}, so you can reuse the same intermediate buffer without having to declare `make([]byte, blockSize)` every single time. You just have to remember to clear it and set the blocksize at the beginning, which I think is a small price to pay compared to declaring `data := make([]byte, blockSize)` every time you call this function.

Reply:
A good point. I have an implementation of this that should reduce slice declaration a lot. Ideally it would read a blockSize worth of bytes directly into the bytes.Buffer but I have not found a way to do that. Let me know if you have any ideas.